optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/uni/DroolsUniCondition.java

```diff
@@ -76,41 +81,69 @@ private DroolsUniCondition(DroolsMetadata<Object, A> aMetadata) {
     }
 
     public DroolsUniCondition<A> andFilter(Predicate<A> predicate) {
-        PatternDSL.PatternDef<Object> newPattern = aMetadata.getPattern()
+        Supplier<PatternDSL.PatternDef<Object>> patternSupplier = () -> aMetadata.buildPattern()
                 .expr(a -> predicate.test(aMetadata.extract(a)));
-        if (aMetadata instanceof DroolsInferredMetadata) {
-            return new DroolsUniCondition<>(((DroolsInferredMetadata) aMetadata).substitute(newPattern));
-        } else {
-            return new DroolsUniCondition<>(((DroolsGenuineMetadata) aMetadata).substitute(newPattern));
-        }
+        return new DroolsUniCondition<>(aMetadata.substitute(patternSupplier));
     }
 
     public <B> DroolsBiCondition<A, B> andJoin(DroolsUniCondition<B> bCondition, AbstractBiJoiner<A, B> biJoiner) {
         DroolsMetadata<Object, B> bMetadata = bCondition.aMetadata;
-        PatternDSL.PatternDef<Object> newPattern = bMetadata.getPattern()
+        Supplier<PatternDSL.PatternDef<Object>> patternSupplier = () -> bMetadata.buildPattern()
                 .expr(aMetadata.getVariableDeclaration(),
                         (b, a) -> matches(biJoiner, aMetadata.extract(a), bMetadata.extract(b)));
-        if (bMetadata instanceof DroolsInferredMetadata) {
-            return new DroolsBiCondition<>(aMetadata, ((DroolsInferredMetadata) bMetadata).substitute(newPattern));
-        } else {
-            return new DroolsBiCondition<>(aMetadata, ((DroolsGenuineMetadata) bMetadata).substitute(newPattern));
-        }
+        return new DroolsBiCondition<>(aMetadata, bMetadata.substitute(patternSupplier));
+    }
+
+    public <ResultContainer, NewA, NewB> List<RuleItemBuilder<?>> completeWithLogicalInsert(
+            Object ruleId, Function<A, NewA> groupKeyMapping,
+            UniConstraintCollector<A, ResultContainer, NewB> collector) {
+        Function1<Object, A> extractor = o -> getAMetadata().extract(o);
+        Function1<Object, NewA> grouper = a -> groupKeyMapping.apply(extractor.apply(a));
+        // Accumulate all NewA into a set.
+        Declaration<NewA> innerNewADeclaration = (Declaration<NewA>) declarationOf(Object.class);
```

Comment:
for bluejeans: why Object?

Reply:
The reason is the same in both cases - `aMetadata` may either be inferred or genuine. So the `extract(a)` either needs to return `a` for genuine or `a.getItem(...)` for inferred. And when we call the function, we do not know which is which.